### PR TITLE
Add support for CoordinateTransformations transforms

### PR DESCRIPTION
**Pull request by SomTambe**
Through this PR, I have added support for the following -

- `LinearMap`, `AffineMap` and `RotMatrix` transforms are now differentiable.
- I modified the `Homography` transform to make it on the lines of `CoordinateTransformations.LinearMap`. Now it is differentiable without the rule I had written.
- I added tests for all of the following, to make sure everything is consistent.

Some things that are breaking - 

- If I use a `AffineMap` combined with a `RotMatrix`, something on the lines of 
```julia
recenter(RotMatrix(π/8), center(img))
```
the gradients come in a nested form compared to the ones when I take using an `AffineMap` composed of purely `SArray`s.
They are something like this - 
```julia
(linear = (mat = (data = [37.49958809560448 -20.572948048408485; 35.68484930615324 -28.041924714099725],),), translation = (data = [1.1544357177976206, -2.9696789827699077],)) # used RotMatrix

(linear = [37.49958809560448 -20.572948048408485; 35.68484930615324 -28.041924714099725], translation = [1.1544357177976206, -2.9696789827699077]) # used StaticArrays
```
I should fix this in another PR.
This PR should solve #17 .

### CODE REVIEW

File: src/geometry/adjoints.jl [hunk 1]
```diff
@@ -50,6 +50,24 @@ function ChainRulesCore.rrule(::Type{SVector{N, T}}, x...) where {N, T}
     return y, svector_const_pb
 end
 
+function ChainRulesCore.rrule(::Type{SMatrix{N, M, T, K}}, x...) where {N, M, T, K}
```
Comment (johnnychen94) [inline src/geometry/adjoints.jl:53]: I believe this method can be absorbed into the previous `SVector` version via something like

```julia
function ChainRulesCore.rrule(::Type{AT}, x...) where {AT<:SArray}
    y = AT(x...)
    ...
end
```

**Review by johnnychen94 — COMMENTED** [review]
With the test passing, nothing too suspicious to me.

> the gradients come in a nested form compared to the ones when I take using an AffineMap composed of purely SArrays.

I believe this is expected? If you try to get gradients for any neural network, you get something similar. @DhairyaLGandhi should know more about how to process this. I guess it's time to introduce Functors.jl to handle nested structures.

File: src/geometry/warp.jl [hunk 2]
```diff
@@ -32,18 +32,21 @@ function Homography{T}() where T
...
+    out = out[1:2] / out[3]
+    return SVector{2, K}(out)
 end
 
-function (h::Homography{T})(x::SVector{2,K}) where {T,K}
-    return h(SVector{3,K}(x...,1))
+function (h::Homography{K})(x::AbstractVector{L}) where {L, K}
+    x = SVector(x...)
+    return h(x)
 end
```
Comment (johnnychen94) [inline src/geometry/warp.jl:45]: Actually, I don't think it's type stable here, you might need to restrict it with length-3 vector with some eager check `length(x) == 3`.

```julia
julia> to_svector(x::AbstractVector) = SVector(x...)

julia> @inferred to_svector([1,2])
ERROR: return type SVector{2, Int64} does not match inferred return type Any
Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] top-level scope
   @ REPL[4]:1

julia> to_svector(x::AbstractVector) = SVector(x...)::SVector{3, eltype(x)}
to_svector (generic function with 1 method)

julia> @inferred to_svector([1,2,3])
3-element SVector{3, Int64} with indices SOneTo(3):
```

File: src/geometry/adjoints.jl [hunk 3]
```diff
@@ -87,14 +89,15 @@ function ChainRulesCore.rrule(::typeof(ImageTransformations.warp!), out, img::Ab
...
+        flds = fieldnames(typeof(tform))
+        ∇tform = canonicalize(Tangent{typeof(tform)}())
         Δy = collect(Δy)
         for p in CartesianIndices(out)
             _, ∇τ = rrule(ImageTransformations._getindex, img, p.I)
             _, _, ∇τ = ∇τ(Δy[p])
             _, ∇ϕ = rrule_via_ad(Zygote.ZygoteRuleConfig(), tform, SVector(p.I))
             ∇h, _ = ∇ϕ(∇τ)
-            ∇tform += Tangent{DiffImages.Homography}(H = ∇h.H)
+            ∇tform += Tangent{typeof(tform)}(;NamedTuple{flds}(∇h)...)
```
Comment (DhairyaLGandhi) [inline src/geometry/adjoints.jl:100]:
```suggestion
            ∇tform += Tangent{typeof(tform)}(NamedTuple{flds}(∇h)...)
```

File: src/geometry/adjoints.jl [hunk 4]
```diff
@@ -50,6 +50,24 @@ function ChainRulesCore.rrule(::Type{SVector{N, T}}, x...) where {N, T}
...
+function ChainRulesCore.rrule(::Type{SMatrix{N, M, T, K}}, x...) where {N, M, T, K}
+    y = SMatrix{N, M, T, K}(x...)
+    function smat_const_pb(Δy)
+        Δy = map(t->eltype(x...)(t...), Δy)
+        return NoTangent(), Δy
+    end
+    return y, smat_const_pb
+end
+
+function ChainRulesCore.rrule(p::Type{RotMatrix{N, T, L}}, x) where {N, T, L}
```
Comment (DhairyaLGandhi) [inline src/geometry/adjoints.jl:62]: Seems like this would be hard to scale to other kinds of transforms if we were to write rules for every transform individually. Are there a countable few of these? Do all of them need special handling or can we AD most? Do these compose to produce other transforms?

**Review by DhairyaLGandhi — COMMENTED** [review]
Yes seems expected. Do we need functors here? Is it even correct to use Functors here?

Btw, very happy to see the rrule of homography be deleted, and consistent code forming. Good job on this PR!

**Comment by DhairyaLGandhi** [comment]
Use  Optimisers.jl along with FluxML/Functors.jl#1 to optimise this. We can do a quick call and I can walk you through structural optimisation.

File: src/geometry/adjoints.jl (hunk 3 above)
Comment (SomTambe) [reply src/geometry/adjoints.jl:100]: Can't do this xD. I wanted to pass the field names too, so have to pass it as kwargs :p
If I splat the namedtuple directly, the values get passed but not the keys.

File: src/geometry/adjoints.jl (hunk 4 above)
Comment (SomTambe) [reply src/geometry/adjoints.jl:62]: Most transforms can be written directly using an `AffineMap`. Others, we can use some existing transform to make that transform. This means we (hopefully) won't have to write any more rules.
We can AD through most, and yes I think we can compose to produce other transforms too. 
I was planning the composing thing in the next PR.

File: src/geometry/adjoints.jl (hunk 1 above)
Comment (SomTambe) [reply src/geometry/adjoints.jl:53]: Yes that seems to be a good idea. I'll make it to the type `::Type{SArray{...}}` only as all `SVector`s and `SMatrix`es can be represented using `SArray`. We can represent higher dimensional Arrays too using the `SArray` interface.

Comment (DhairyaLGandhi) [reply src/geometry/adjoints.jl:53]: What happens if you nograd this kind of rrule entirely?

File: src/geometry/adjoints.jl (hunk 4 above)
Comment (DhairyaLGandhi) [reply src/geometry/adjoints.jl:62]: It would be good to have a poc that demos what you're talking about.

File: test/geometry/adjoints.jl [hunk 5]
```diff
@@ -53,25 +53,56 @@ end
...
     for t in (Float32, Float64, RGB{Float32}, RGB{Float64})
         inp = rand(t, 2)
+        inp_mat = rand(t, 3, 3)
         _sep(x) = x.r + x.g + x.b
         if t ∈ (Float32, Float64)
             @test Zygote.gradient(x->sum(SVector{2, t}(x)), inp)[1] == ones(t, 2)
+            @test Zygote.gradient(x->sum(SMatrix{3, 3, t, 9}(x)), inp_mat)[1] == ones(t, 3, 3)
         else
             @test Zygote.gradient(x->_sep(sum(SVector{2, t}(x))), inp)[1] == ones(t, 2)
+            @test Zygote.gradient(x->_sep(sum(SMatrix{3, 3, t, 9}(x))), inp_mat)[1] == ones(t, 3, 3)
```
Comment (DhairyaLGandhi) [inline test/geometry/adjoints.jl:66]: The tests in both the branches of the if blocks seems the same modulo `_sep`, should we get rid of the if block perhaps

File: test/geometry/adjoints.jl [hunk 6]
```diff
@@ -53,25 +53,56 @@ end
...
     end
 end
 
 @testset "DiffImages.Homography method gradient" begin
-    h = DiffImages.Homography()
+    h = DiffImages.Homography{Float64}()
     v = rand(SVector{2, Float64})
     zy = Zygote.gradient((x,y)->sum(y(x)), v, h)
     fd = grad(central_fdm(5,1), (x,y)->sum(y(x)), v, h)
-    @test zy[1] ≈ fd[1]
```
Comment (DhairyaLGandhi) [inline test/geometry/adjoints.jl:73]: Why are we getting rid of the test?

File: src/geometry/adjoints.jl (hunk 1 above)
Comment (SomTambe) [reply src/geometry/adjoints.jl:53]: That wouldn't be a good idea. We use StaticArrays widely in intermediate representation, and that would surely cause gradients to stop propagating.
Let me try that once.

File: test/geometry/warp.jl [hunk 7]
```diff
@@ -84,3 +84,27 @@ end
...
+    f(t) = begin
+        out = ImageTransformations.warp(img, t, axes(img), zero(eltype(img)))
+        loss = image_mse(out, tgt)
+        loss
+    end
+
+    tfm1 = recenter(RotMatrix(π/8), center(img)) # using RotMatrix
+    tfm2 = AffineMap(SMatrix{2, 2, Float64, 4}([cos(π/8) -sin(π/8); sin(π/8) cos(π/8)]), tfm1.translation)
+
+    zy1 = Zygote.gradient(f, tfm1)
```
Comment (DhairyaLGandhi) [inline test/geometry/warp.jl:101]: Oh brilliant, are the `warp!` rules working as expected now?

File: src/geometry/adjoints.jl (hunk 4 above)
Comment (SomTambe) [reply src/geometry/adjoints.jl:62]: Yes I'll prepare it with the docs.

File: test/geometry/adjoints.jl (hunk 6 above)
Comment (SomTambe) [reply test/geometry/adjoints.jl:73]: It got removed accidentally. Thanks for reminding me, I have added a better version of the same now.

File: test/geometry/adjoints.jl (hunk 5 above)
Comment (SomTambe) [reply test/geometry/adjoints.jl:66]: This seems like a good idea. I'll make this consistent by placing an `abs` equivalent.

File: test/geometry/warp.jl (hunk 7 above)
Comment (SomTambe) [reply test/geometry/warp.jl:101]: Yup

File: src/geometry/warp.jl [hunk 8]
```diff
@@ -40,7 +40,7 @@ function (h::Homography{K})(x::SVector{3, L}) where {L, K}
...
 
 function (h::Homography{K})(x::AbstractVector{L}) where {L, K}
-    x = SVector(x...)
+    x = SVector(x...)::SMatrix{3, 3, L, 9}
```
Comment (johnnychen94) [inline src/geometry/warp.jl:43]: Are you sure this ever works? Is this method covered by the test?

Comment (SomTambe) [reply src/geometry/warp.jl]: Sorry for the goof-up, should not have been this. I have rectified it and added tests for the same.

**Comment by SomTambe** [comment]
Since I have made the required changes, can we get this PR merged? 
About the functor things, I can merge the related changes in a separate PR. 
We can have a short call as Dhairya had suggested. I have been a bit busy preparing the docs and due to academic commitments.

File: Project.toml [hunk 9]
```diff
@@ -13,6 +13,7 @@ ImageCore = "a09fc81d-aa75-5fe9-8630-4744c3626534"
 ImageTransformations = "02fcd773-0e25-5acc-982a-7f6622650795"
 Interpolations = "a98d9a8b-a2ab-59e6-89dd-64a1c18fca59"
 LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
+Rotations = "6038ab10-8711-5258-84ad-4b1120ba62dc"
```
Comment (johnnychen94) [inline Project.toml:16]: Generally, every package in [deps] should have associated entry in [compat].

**Review by johnnychen94 — APPROVED** [review]
Looks good to me. I'll leave it to @DhairyaLGandhi since there seems one or two unresolved comments.
